### PR TITLE
Override app namespace

### PR DESCRIPTION
**Pull request by stinkyfingers**
# Description

Permits overriding an App's namespace. Apps default to the namespace specified by `framework.spec.namespace`, but there are cases where we don't want namespace tied to framework.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [ ] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [ ] All added public packages, funcs, and types have been documented with doc comments
- [ ] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [ ] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings

## Additional Information (omit if empty)

Please include anything else relevant to this PR that may be useful to know.

### CODE REVIEW

**Review by aleksej-paschenko — CHANGES_REQUESTED** [review]

File: internal/chart/testdata/render_yamls/annotations-postrender.yaml [hunk 1]
```diff
@@ -46,7 +46,7 @@ kind: Deployment
 metadata:
   annotations:
     app: app1
-    theketch.io/is-shipa: "true"
+    shipa.io/is-shipa: "true"
```
[the remark this answers is not in the document]
Comment (stinkyfingers) [reply internal/chart/testdata/render_yamls/annotations-postrender.yaml]: Looks like a rebase got this caught up.

File: internal/controllers/app_controller.go [hunk 2]
```diff
@@ -828,6 +847,33 @@ func (r *AppReconciler) deleteChart(ctx context.Context, app *ketchv1.App) error
...
+	namespacedName := framework.Spec.NamespaceName
+	if app.Spec.Namespace != "" { // override framework.spec.namespace if app.spec.namespace is set
+		namespacedName = app.Spec.Namespace
+	}
+	var namespace v1.Namespace
+	err = r.Get(ctx, types.NamespacedName{Name: namespacedName}, &namespace)
+	if err != nil {
+		return err
+	}
+	namespace.Labels["istio-injection"] = "enabled"
```
[the remark this answers is not in the document]
Comment (stinkyfingers) [reply internal/controllers/app_controller.go:868]: Modified this function to remove this `istio-injection` label from the `namespace` when not istio. Also we now reconcile the `sidecar.istio.io/inject` label on Pods.

File: internal/controllers/app_controller.go [hunk 3]
```diff
@@ -828,6 +847,34 @@ func (r *AppReconciler) deleteChart(ctx context.Context, app *ketchv1.App) error
...
+	if framework.Spec.IngressController.IngressType != ketchv1.IstioIngressControllerType {
+		delete(namespace.Labels, "istio-injection")
+		app.AddLabel(map[string]string{"sidecar.istio.io/inject": "false"}, ketchv1.Target{Kind: "Pod", APIVersion: "v1"})
```
Comment (aleksej-paschenko) [inline internal/controllers/app_controller.go:871]: with `"sidecar.istio.io/inject": "false"` label should we just keep a namespace as it is? 
meaning we really don't need to care about a namespace's labels.

File: internal/api/v1beta1/app_types.go [hunk 4]
```diff
@@ -664,6 +667,34 @@ func (app *App) DoRollback() {
...
 }
 
+// AddLabel adds a label to an app's deployments' processes. It will remove labels with matching keys and targets.
+func (app *App) AddLabel(label map[string]string, target Target) {
+	// clean up labels
+	for i := len(app.Spec.Labels) - 1; i >= 0; i-- {
+		if app.Spec.Labels[i].Target != target {
+			continue
+		}
+		for key := range app.Spec.Labels[i].Apply {
```
Comment (aleksej-paschenko) [inline internal/api/v1beta1/app_types.go:677]: that's a very tricky function
should it be 
```golang
for key := range labels {
    delete(app.Spec.Labels[i].Apply, key)
}
```
?

File: internal/controllers/app_controller.go (hunk 3 above)
Comment (stinkyfingers) [reply internal/controllers/app_controller.go:860]: Sounds good. I removed all the namespace label changes.

File: internal/api/v1beta1/app_types.go (hunk 4 above)
Comment (stinkyfingers) [reply internal/api/v1beta1/app_types.go]: Good catch. Changed.

**Comment by stinkyfingers** [comment]
I opted to update this to include a `namespace` override for `job` too.

**Review by aleksej-paschenko — APPROVED** [review]

**Pull request merged**